src/io/asyncsocket.c

```diff
@@ -673,6 +673,176 @@ MVMObject * MVM_io_socket_connect_async(MVMThreadContext *tc, MVMObject *queue,
+    UnixConnectInfo *ci = (UnixConnectInfo *)data;
+    ci->tc        = tc;
+    ci->work_idx  = MVM_io_eventloop_add_active_work(tc, async_task);
+
+    /* Create and initialize socket and connection. */
+    ci->socket        = MVM_malloc(sizeof(uv_tcp_t));
+    ci->connect       = MVM_malloc(sizeof(uv_connect_t));
+    ci->connect->data = data;
+    if ((r = uv_pipe_init(loop, ci->socket, 1)) < 0) {
+        uv_pipe_connect(ci->connect, ci->socket, ci->dest, on_unix_connect);
+        /* Error; need to notify. */
+        MVMROOT(tc, async_task, {
+            MVMObject    *arr = MVM_repr_alloc_init(tc, tc->instance->boot_types.BOOTArray);
+            MVM_repr_push_o(tc, arr, ((MVMAsyncTask *)async_task)->body.schedulee);
+            MVM_repr_push_o(tc, arr, tc->instance->boot_types.BOOTIO);
+            MVMROOT(tc, arr, {
+                MVMString *msg_str = MVM_string_ascii_decode_nt(tc,
+                    tc->instance->VMString, uv_strerror(r));
+                MVMObject *msg_box = MVM_repr_box_str(tc,
+                    tc->instance->boot_types.BOOTStr, msg_str);
+                MVM_repr_push_o(tc, arr, msg_box);
+                MVM_repr_push_o(tc, arr, tc->instance->boot_types.BOOTStr);
+                MVM_repr_push_o(tc, arr, tc->instance->boot_types.BOOTInt);
+                MVM_repr_push_o(tc, arr, tc->instance->boot_types.BOOTStr);
+                MVM_repr_push_o(tc, arr, tc->instance->boot_types.BOOTInt);
+            });
+            MVM_repr_push_o(tc, ((MVMAsyncTask *)async_task)->body.queue, arr);
+        });
+
+        /* Cleanup handles. */
+        MVM_free_null(ci->connect);
+        uv_close((uv_handle_t *)ci->socket, free_on_close_cb);
+        ci->socket = NULL;
+        MVM_io_eventloop_remove_active_work(tc, &(ci->work_idx));
+    }
+}
+
+/* Frees info for a connection task. */
+static void unix_connect_gc_free(MVMThreadContext *tc, MVMObject *t, void *data) {
+    if (data) {
+        UnixConnectInfo *ci = (UnixConnectInfo *)data;
+        if (ci->dest)
+            MVM_free(ci->dest);
+        MVM_free(ci);
+    }
+}
+
+/* Operations table for async connect task. */
+static const MVMAsyncTaskOps unix_connect_op_table = {
+    unix_connect_setup,
+    NULL,
+    NULL,
+    NULL,
+    unix_connect_gc_free
+};
+/* Sets off an asynchronous socket connection. */
+MVMObject * MVM_io_socket_connect_unix_async(MVMThreadContext *tc, MVMObject *queue,
+                                        MVMObject *schedulee, MVMString *path,
+                                        MVMObject *async_type) {
+    MVMAsyncTask *task;
+    UnixConnectInfo  *ci;
+
+    /* Validate REPRs. */
+    if (REPR(queue)->ID != MVM_REPR_ID_ConcBlockingQueue)
+        MVM_exception_throw_adhoc(tc,
+            "asyncconnect target queue must have ConcBlockingQueue REPR");
+    if (REPR(async_type)->ID != MVM_REPR_ID_MVMAsyncTask)
+        MVM_exception_throw_adhoc(tc,
+            "asyncconnect result type must have REPR AsyncTask");
```

Comment:
You can use `expected_reprs` in the syscall definition to enforce these rather than having to do it imperatively here (that will also let spesh eliminate the type checks if this is a hot path).

Reply:
That sounds sensible